ZEP008-hot-backup.md

```diff
@@ -0,0 +1,619 @@
+On reading a command:
+```
+if state.checkpointId < command.checkpointId
+    send a command `takeBackup checkpointId command.position` to BackupActor        
+    state.checkpointId = command.checkpointId
+    state.checkpointPosition = command.checkpointPosition
+    write follow up event, `checkpointTaken checkpointId`
+Process command
+    If an engine command, engine will process it.
+    If it is `checkpoint` record, then send a response to the coordinator.
+    If it is `restore X` record, then
+        update state.checkpointId=X
+        update state.checkpointPosition=command.position
+        write followup event `restored X`
+```
+During Replay:
+```
+On replaying `checkpointTaken X` event,
+    update state.checkpointId=X
+    update state.checkpointPosition=event.sourcePosition.
+On replaying `restored X`
+    update state.checkpointId=X
+    update state.checkpointPosition=event.sourcePosition
+```
+Inter-partition communication:  
+- Any command that is sent from a StreamProcessor to another partition's StreamProcessor will contain `state.checkpointId` at the time the command was created by the sender. If the command has to be resent, it uses the same checkpointId used in the first attempt.
+
+
+###### SnapshotStoreActor:
+```
+On snapshot lock request
+	Mark the snapshot as locked in memory
+	A locked snapshot should not be deleted until it is unlocked
+	When a snapshot is locked, the corresponding logs should be not compacted as well
+On snapshot unlock request
+	Remove the in-memory lock
+	Delete the snapshot if there is a newer snapshot
+	Compact the logs
+```
+###### SnapshotDirector:
+Before committing the snapshot check the following:
+If `snapshotId.processedPosition < streamProcessor.lastCheckpointPosition < lastWrittenPosition`, then abort the snapshot. This is needed to prevent the scenario where the concurrently taken snapshot has `processedPosition` > `checkpoint position`.
+
+#### On Restore
+When restoring from a backup
+- Delete all records with position > checkpointPosition (including the command at the checkpointPosition.)
+- Write new record `Restore backupId` at the checkpointPosition
+
+Note:- Restore process should be completed before any normal operation of Zeebe including raft leader election. No new records should be written to the log before restore process is completed.
+
+### Edge cases and failure scenarios
+
+Explain known edge cases and expected failure scenarios and describe how the above algorithm handles them.
+
+##### Edge cases due to concurrent snapshots and compaction
+
+While taking the backup, the available snapshot has `processedPosition >= checkpointPosition`. This would mean that we cannot take a valid backup, because there is not way to retrieve a state that represents the checkpoint until the checkpoint command.
+
+In this there are two scenarios that can happen:
+
+1. The available snapshot is after the `checkpointPosition`. That s`napshotId.processedPosition > checkpointPosition`.
```

Comment:
The problem is backup is taken asynchronously. StreamProcessor triggers the backup by sending a command to the BackupActor. StreamProcessor continues processing immediately after that. So by the time BackupActor starts taking the backup, the processed position would have advanced. A new snapshot might have been created with the advanced processed position and the log is already compacted. The chances of this happening might be less, but in theory this can happen.